core/Tracker.php

```diff
@@ -109,6 +110,16 @@ public function main(Handler $handler, RequestSet $requestSet)
             $this->init();
             $handler->init($this, $requestSet);
 
+            // if the tracker config is requested, don't bother trying to track anything
```

Comment:
btw one thing that might be interesting looking at is similar behaviour to `Configs` class in `getCachedWebsiteAttributes` method in HeatmapSessionRecording where we basically try to read the website specific tracker cache and if it exists, avoid loading plugins. We might be able to later add some small performance tweak in matomo.php to behave similar as in `configs.php` that we directly get `$_GET['idsite'];` when the `configs=1` parameter is set, and then before even doing anything tracker specific only check if the website cache exists.  This could also have the benefit that it would maybe work quite well with QueuedTracking etc. To be seen though and something we can add later

Reply:
Would this would be a problem for a plugin that didn't want to use the tracker cache, for example, getting the next page to scan?

Reply:
> getting the next page to scan

Not sure what you mean here? It's eventually cached in the browser anyway?

Reply:
The client side cache isn't important for this use case, since the scan would happen once at the start of the visit and then, some other visit would scan another page. But server side, we would not cache the "next page to scan", since it would have to be fetched each time. If we check for whether the website specific tracker cache exists and avoid loading the tracker plugins, then we won't be able to execute the event and the plugin won't be able to return the right data, correct? Ie, some plugins may not want to use the cache. I'm not sure I'm understanding the proposed optimization though, it is to not have to load tracker plugins and execute the event server side, right?

Reply:
Not quite getting it yet re next page scan but the idea is basically, since this request would be executed for each visit at least once (and many visits have only one page view), that we try to make this as efficient as possible - if possible. Basically, it would mean moving more decision making into the client when possible to save a lot of resources etc.

Reply:
Does that mean disabling tracker plugins for some requests and only using the cache? That's the issue I'm thinking of since it will mean the event handler for a plugin won't be executed, but that would be an issue for a plugin that doesn't necessarily use the cache (unless I'm not understanding).

I think we could still do that optimization though, it would require having plugins provide information about what cache entries they add, and under what conditions it's ok to not load the plugin. Then we could avoid loading individual plugins if everything is available in the cache.

Reply:
> Does that mean disabling tracker plugins for some requests and only using the cache?

The idea is, check if the website tracker cache exists and if so, read the result from the cache.

If cache does not exist, we would launch the tracker mode regularly etc and fetch the tracker cache from the plugins which also add the information for configs etc. 

Not sure if that makes it bit more clear?

Reply:
That is what I understood the idea to be, but it poses a problem for plugins that don't use the cache, since even if the cache exists, there will be no value to read.

Reply:
>  but it poses a problem for plugins that don't use the cache, since even if the cache exists, there will be no value to read.

Sorry not getting this part. Not sure which cache or plugins you mean etc

Reply:
So let's say a plugin uses this event to:

```
function getTrackerConfigs(&$configs)
{
    $configs['MyPlugin']['nextPageToScan'] = $this->getNextPageToScan(); // ... performs a query on the DB ...
}
```

then the Tracker checks for a cache entry:

```
$cache = Cache::getCacheWebsiteAttributes($idSite);
if (!empty($cache['trackerconfigs'])) {
    return $cache['trackerconfigs']; // in this case, `$configs['MyPlugin']['nextPageToScan']` is either absent because we don't cache it, or incorrect and old (if we cache the result of the event)
}

// ... load tracker plugins ...
Piwik::postEvent('Tracker.getTrackerConfigs', $configs);
return $configs;
```

Since the value used by the plugin is dynamic it can't be cached. For this optimization, we'd have to do something like:

```
$pluginsToLoad = [];
foreach ($trackerPlugins as $plugin) {
    if (empty($cache['trackerconfigs'][$plugin]) && isPluginUsingTrackerCacheForConfigs($plugin)) {
        $pluginsToLoad[] = $plugin;
    }
}

// load $pluginsToLoad
Piwik::postEvent('Tracker.getTrackerConfigs', $configs);
$cache['trackerconfigs'] = array_merge($cache['trackerconfigs'], getConfigsForPluginsUsingCache($configs));
```

I hope that makes sense. This isn't for an existing plugin, btw.

Reply:
Got it now 👍 And we clearly wouldn't want to return all `nextPagesToScan ` as it could be a long list etc.

Reply:
@tsteur I added the optimization for this, I haven't tested it since I'm not completely setup for it, but it should be possible to avoid loading some plugins when it's called if data is cached.